ddl/partition.go

```diff
@@ -365,6 +376,52 @@ func onDropTablePartition(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 	return ver, nil
 }
 
+// onDropTablePartition truncates old partition meta.
+func onTruncateTablePartition(t *meta.Meta, job *model.Job) (int64, error) {
+	var ver int64
+	var partName string
+	if err := job.DecodeArgs(&partName); err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+	tblInfo, err := getTableInfo(t, job, job.SchemaID)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	pi := tblInfo.GetPartitionInfo()
+	if pi == nil {
+		return ver, errors.Trace(ErrPartitionMgmtOnNonpartitioned)
+	}
+
+	oldID := int64(-1)
+	for i := 0; i < len(pi.Definitions); i++ {
+		def := &pi.Definitions[i]
+		if strings.EqualFold(def.Name.L, strings.ToLower(partName)) {
+			pid, err1 := t.GenGlobalID()
+			if err != nil {
+				return ver, errors.Trace(err1)
+			}
+			oldID = def.ID
+			def.ID = pid
+			break
+		}
+	}
+	if oldID == -1 {
+		return ver, errUnknownPartition.GenWithStackByArgs(partName, tblInfo.Name.O)
+	}
+
+	ver, err = updateVersionAndTableInfo(t, job, tblInfo, true)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	// Finish this job.
+	job.FinishTableJob(model.JobStateDone, model.StateNone, ver, tblInfo)
+	// A background job will be created to delete old partition data.
```

Comment:
You should change `delete_range` to handle `ActionTruncateTablePartition` job. Otherwise, the data of old partition won't be clean.

Reply:
done